File: searching_algo/binary_search.c
// Binary Search
// 이진 탐색

#include <stdio.h>
#include <time.h>

#define MAX_ELEMENTS 10000000L

int list[MAX_ELEMENTS];
int count = 0;

// make list
int make_list(int first, int last) {

	int index = 0;
	
	for (int i = first; i <= last; i++) {
		list[index] = i;
		index++;
	}

	return 0;
}

// binary search algo
int search_binary(int key, int low, int high) {

	int middle;

	if (low <= high) {
		middle = (low + high) / 2;

		if (key == list[middle]) {
			count++;
			printf("Search Count = %d", count);
			return middle;

		} else if (key < list[middle]) {
			count++;
			return search_binary(key, low, middle - 1);

		} else {
			count++;
			return search_binary(key, middle + 1, high);

		}
	}

	return -1;
}

int main() {
	make_list(0, 10);
	search_binary(4, 0, 10);
}

File: searching_algo/binary_search.py
# Binary Search

def binary_search(num_list, key):

    num_list.sort()

    start = 0
    last = len(num_list) - 1
    count = 0

    while start <= last:
        middle = (start + last) // 2

        if num_list[middle] == key:
            count += 1
            print("Search Count =", count)
            return middle

        elif num_list[middle] < key:
            start = middle + 1
            count += 1

        else:
            last = middle - 1
            count += 1

    return -1

num_list = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
binary_search(num_list, 4)
